jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/ResultAttributes.java

```diff
@@ -0,0 +1,164 @@
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Enumerates the attributes of {@link Result}
+ *
+ * @author Simon Hutchinson
+ *         <a href="https://github.com/sihutch">github.com/sihutch</a>
+ */
+public final class ResultAttributes {
+	private ResultAttributes() {
+	}
+
+	public static final ResultAttribute TYPE_NAME = new ResultAttribute("typeName") {
+		@Override
+		public String getAttribute(Result result) {
+			return result.getTypeName();
+		}
+	};
+
+	public static final ResultAttribute OBJ_DOMAIN = new ResultAttribute("objDomain") {
+		@Override
+		public String getAttribute(Result result) {
+			return result.getObjDomain();
+		}
+	};
+
+	public static final ResultAttribute CLASS_NAME = new ResultAttribute("className") {
+		@Override
+		public String getAttribute(Result result) {
+			return result.getClassName();
+		}
+	};
+
+	public static final ResultAttribute ATTRIBUTE_NAME = new ResultAttribute("attributeName") {
+		@Override
+		public String getAttribute(Result result) {
+			return result.getAttributeName();
+		}
+	};
+
+	/**
+	 * Implementation of {@link ResultAttribute} to lookup type name properties
+	 */
+	private static final class TypeNameProperty extends ResultAttribute {
+		private static final String PREFIX = "typeName.";
+		private final String propertyName;
+
+		private TypeNameProperty(String propertyName) {
+			super(PREFIX + propertyName);
+			this.propertyName = propertyName;
+		}
+
+		@Override
+		public String getAttribute(Result result) {
+			return result.getTypeNameMap().get(propertyName);
+		}
+	}
+
+	/**
+	 * Get the {@link ResultAttributes} value from the attribute name
+	 *
+	 * @param attributeName <p>The attribute name for the {@link ResultAttribute} allowed values are:</p>
+	 *                      <ul>
+	 *                      <li>typeName</li>
+	 *                      <li>objDomain</li>
+	 *                      <li>className</li>
+	 *                      <li>attributeName</li>
+	 *                      </ul>
+	 * @return the {@link ResultAttribute}
+	 */
+	public static ResultAttribute fromAttribute(@Nonnull String attributeName) {
+		if (attributeName.startsWith(TypeNameProperty.PREFIX)) {
+			return new TypeNameProperty(attributeName.substring(TypeNameProperty.PREFIX.length()));
+		}
+		String[] split = StringUtils.splitByCharacterTypeCamelCase(attributeName);
+		StringBuilder sb = new StringBuilder(split[0].toUpperCase()).append("_").append(split[1].toUpperCase());
+		return valueOf(sb.toString());
+	}
+
+	/**
+	 * Get the {@link ResultAttributes} value for each attribute name.
+	 *
+	 * @return Set of {@link ResultAttribute}
+	 * @see #fromAttribute(String)
+	 */
+	public static ImmutableSet<ResultAttribute> fromAttributes(@Nonnull Collection<String> attributeNames) {
+		Set<ResultAttribute> set = new HashSet<ResultAttribute>(attributeNames.size());
```

Comment:
Youcould use an [`ImmutableSet.Builder`](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/ImmutableSet.Builder.html) instead of an intermediate `HashSet`.